File: data/dataEngine.ts
/**
 * Deterministic data generation engine for CalmCampus.
 *
 * Generates realistic behavioral data based on the actual date.
 * The same date always produces the same values (seeded by date string),
 * but values vary naturally across days with patterns like:
 *   - Weekday stress (worse sleep, more screen time mid-week)
 *   - Weekend recovery
 *   - Gradual trends over weeks
 *   - Occasional "bad days"
 */

export interface DailyData {
  date: string;
  dayLabel: string;
  dayOfWeek: number; // 0=Sun, 6=Sat
  sleepHours: number;
  screenTimeHours: number;
  typingSpeed: number;
  appSwitches: number;
  wellnessScore: number;
}

export interface BehavioralSignal {
  id: string;
  label: string;
  value: string;
  unit: string;
  trend: 'up' | 'down' | 'stable';
  trendIsGood: boolean;
  icon: string;
}

// Simple seeded pseudo-random number generator (mulberry32)
function seededRandom(seed: number): () => number {
  return () => {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// Convert date string to numeric seed
function dateSeed(dateStr: string): number {
  let hash = 0;
  for (let i = 0; i < dateStr.length; i++) {
    const char = dateStr.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash |= 0;
  }
  return Math.abs(hash);
}

// Clamp a value between min and max
function clamp(val: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, val));
}

// Round to 1 decimal
function round1(val: number): number {
  return Math.round(val * 10) / 10;
}

const DAY_NAMES = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

/**
 * Generate data for a single day based on date.
 * Patterns:
 *   - Mon-Wed: increasing stress (less sleep, more screen time)
 *   - Thu: peak stress day
 *   - Fri: slight recovery
 *   - Sat-Sun: recovery (better sleep, less screen time)
 */
function generateDay(date: Date): DailyData {
  const dateStr = date.toISOString().split('T')[0];
  const rand = seededRandom(dateSeed(dateStr));
  const dow = date.getDay(); // 0=Sun, 6=Sat

  // Day-of-week stress multiplier (0 = no stress, 1 = max stress)
  const stressByDay: Record<number, number> = {
    0: 0.1,  // Sunday - relaxed
    1: 0.3,  // Monday - ramping up
    2: 0.5,  // Tuesday - building
    3: 0.7,  // Wednesday - peak midweek
    4: 0.8,  // Thursday - highest stress
    5: 0.5,  // Friday - winding down
    6: 0.15, // Saturday - relaxed
  };
  const stress = stressByDay[dow];

  // Add some daily randomness (±0.15)
  const noise = (rand() - 0.5) * 0.3;
  const effectiveStress = clamp(stress + noise, 0, 1);

  // Occasional "bad day" (~10% chance)
  const isBadDay = rand() < 0.10;
  const badDayBoost = isBadDay ? 0.3 : 0;

  const totalStress = clamp(effectiveStress + badDayBoost, 0, 1);

  // Generate metrics based on stress level
  // Sleep: 8.5h when relaxed → 4.5h when very stressed
  const sleepBase = 8.5 - totalStress * 4.0;
  const sleepHours = round1(clamp(sleepBase + (rand() - 0.5) * 0.8, 4.0, 9.5));

  // Screen time: 2h when relaxed → 8h when stressed
  const screenBase = 2.5 + totalStress * 5.5;
  const screenTimeHours = round1(clamp(screenBase + (rand() - 0.5) * 1.0, 1.5, 9.0));

  // Typing speed: 55 wpm when relaxed → 30 wpm when stressed
  const typingBase = 55 - totalStress * 25;
  const typingSpeed = Math.round(clamp(typingBase + (rand() - 0.5) * 6, 25, 60));

  // App switches per hour: 8 when focused → 35 when distracted
  const switchBase = 8 + totalStress * 27;
  const appSwitches = Math.round(clamp(switchBase + (rand() - 0.5) * 6, 5, 40));

  // Wellness score: composite (weighted average of normalized metrics)
  const sleepScore = clamp((sleepHours - 4) / 5, 0, 1);         // 4h=0, 9h=1
  const screenScore = clamp(1 - (screenTimeHours - 1.5) / 7, 0, 1); // 1.5h=1, 8.5h=0
  const typingScore = clamp((typingSpeed - 25) / 35, 0, 1);     // 25=0, 60=1
  const focusScore = clamp(1 - (appSwitches - 5) / 30, 0, 1);   // 5=1, 35=0

  const rawWellness = (sleepScore * 0.35 + screenScore * 0.25 + typingScore * 0.15 + focusScore * 0.25) * 100;
  const wellnessScore = Math.round(clamp(rawWellness + (rand() - 0.5) * 6, 10, 98));

  return {
    date: dateStr,
    dayLabel: DAY_NAMES[dow],
    dayOfWeek: dow,
    sleepHours,
    screenTimeHours,
    typingSpeed,
    appSwitches,
    wellnessScore,
  };
}

/**
 * Get data for the last N days (including today).
 */
export function getWeeklyData(days: number = 7): DailyData[] {
  const result: DailyData[] = [];
  const now = new Date();

  for (let i = days - 1; i >= 0; i--) {
    const d = new Date(now);
    d.setDate(d.getDate() - i);
    result.push(generateDay(d));
  }

  return result;
}

/**
 * Get today's data.
 */
export function getTodayData(): DailyData {
  return generateDay(new Date());
}

/**
 * Get today's behavioral signals with trend comparison to yesterday.
 */
export function getTodaySignals(): BehavioralSignal[] {
  const today = getTodayData();
  const yesterday = generateDay((() => { const d = new Date(); d.setDate(d.getDate() - 1); return d; })());

  return [
    {
      id: 'sleep',
      label: 'Sleep',
      value: today.sleepHours.toFixed(1),
      unit: 'hours',
      trend: today.sleepHours > yesterday.sleepHours ? 'up' : today.sleepHours < yesterday.sleepHours ? 'down' : 'stable',
      trendIsGood: today.sleepHours >= 7,
      icon: 'moon',
    },
    {
      id: 'screen',
      label: 'Screen Time',
      value: today.screenTimeHours.toFixed(1),
      unit: 'hours',
      trend: today.screenTimeHours > yesterday.screenTimeHours ? 'up' : today.screenTimeHours < yesterday.screenTimeHours ? 'down' : 'stable',
      trendIsGood: today.screenTimeHours <= 5,
      icon: 'phone',
    },
    {
      id: 'typing',
      label: 'Typing Pattern',
      value: today.typingSpeed.toString(),
      unit: 'wpm',
      trend: today.typingSpeed > yesterday.typingSpeed ? 'up' : today.typingSpeed < yesterday.typingSpeed ? 'down' : 'stable',
      trendIsGood: today.typingSpeed >= 45,
      icon: 'keyboard',
    },
    {
      id: 'focus',
      label: 'Focus Score',
      value: Math.round(100 - (today.appSwitches / 40) * 100).toString(),
      unit: '%',
      trend: today.appSwitches < yesterday.appSwitches ? 'up' : today.appSwitches > yesterday.appSwitches ? 'down' : 'stable',
      trendIsGood: today.appSwitches <= 15,
      icon: 'target',
    },
  ];
}

/**
 * Generate an AI insight based on the weekly data.
 */
export function getWeeklyInsight(): string {
  const week = getWeeklyData(7);
  const today = week[week.length - 1];
  const avgSleep = week.reduce((s, d) => s + d.sleepHours, 0) / week.length;
  const avgScreen = week.reduce((s, d) => s + d.screenTimeHours, 0) / week.length;
  const worstDay = week.reduce((worst, d) => d.wellnessScore < worst.wellnessScore ? d : worst, week[0]);
  const bestDay = week.reduce((best, d) => d.wellnessScore > best.wellnessScore ? d : best, week[0]);
  const trend = today.wellnessScore - week[0].wellnessScore;

  if (today.sleepHours < 6) {
    return `You slept only ${today.sleepHours}h last night. Sleep deprivation compounds — even one recovery night helps. Consider winding down earlier tonight.`;
  }
  if (today.screenTimeHours > 6) {
    return `Screen time is elevated at ${today.screenTimeHours}h today. Try the 20-20-20 rule: every 20 minutes, look at something 20 feet away for 20 seconds.`;
  }
  if (worstDay.wellnessScore < 50) {
    return `${worstDay.dayLabel} was your toughest day this week (score: ${worstDay.wellnessScore}). Your best was ${bestDay.dayLabel} at ${bestDay.wellnessScore}. The recovery pattern looks healthy.`;
  }
  if (avgSleep < 6.5) {
    return `Your average sleep this week is ${avgSleep.toFixed(1)}h — below the recommended 7-9h. Sleep is the strongest predictor of next-day wellbeing in students.`;
  }
  if (trend > 10) {
    return `Your wellness trend is improving (+${trend} points since ${week[0].dayLabel}). Keep maintaining your current routine — consistency is key.`;
  }
  if (avgScreen > 5.5) {
    return `Average screen time is ${avgScreen.toFixed(1)}h this week. Consider setting app timers for your most-used apps to stay mindful of usage.`;
  }
  return `Your wellness has been ${today.wellnessScore >= 70 ? 'steady' : 'variable'} this week. ${bestDay.dayLabel} was your strongest day. Keep prioritizing sleep and breaks.`;
}

File: lib/data/data_engine.dart
import '../models/daily_data.dart';
import '../models/behavioral_signal.dart';
import '../models/wellness_anomaly.dart';

const _dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

// Mulberry32 seeded PRNG
typedef _Rng = double Function();

_Rng _seededRandom(int seed) {
  int s = seed;
  return () {
    s |= 0;
    s = (s + 0x6D2B79F5) | 0;
    int t = (s ^ (s >> 15)) * (1 | s);
    t = (t + (t ^ (t >> 7)) * (61 | t)) ^ t;
    return ((t ^ (t >> 14)) & 0x7FFFFFFF) / 0x7FFFFFFF;
  };
}

int _dateSeed(String dateStr) {
  int hash = 0;
  for (int i = 0; i < dateStr.length; i++) {
    hash = ((hash << 5) - hash) + dateStr.codeUnitAt(i);
    hash &= 0x7FFFFFFF;
  }
  return hash.abs();
}

double _clamp(double val, double lo, double hi) => val.clamp(lo, hi);
double _round1(double val) => (val * 10).round() / 10;

DailyData _generateDay(DateTime date) {
  final dateStr =
      '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  final rand = _seededRandom(_dateSeed(dateStr));
  final dow = date.weekday % 7; // 0=Sun in JS, Dart weekday 7=Sun

  const stressByDay = <int, double>{
    0: 0.1, // Sun
    1: 0.3, // Mon
    2: 0.5, // Tue
    3: 0.7, // Wed
    4: 0.8, // Thu
    5: 0.5, // Fri
    6: 0.15, // Sat
  };
  final stress = stressByDay[dow] ?? 0.5;
  final noise = (rand() - 0.5) * 0.3;
  final effectiveStress = _clamp(stress + noise, 0, 1);
  final isBadDay = rand() < 0.10;
  final totalStress = _clamp(effectiveStress + (isBadDay ? 0.3 : 0), 0, 1);

  final sleepBase = 8.5 - totalStress * 4.0;
  final sleepHours = _round1(_clamp(sleepBase + (rand() - 0.5) * 0.8, 4.0, 9.5));

  final screenBase = 2.5 + totalStress * 5.5;
  final screenTimeHours = _round1(_clamp(screenBase + (rand() - 0.5) * 1.0, 1.5, 9.0));

  // Active minutes: higher stress → less activity
  final activityBase = 50 - totalStress * 40;
  final activeMinutes = _clamp(activityBase + (rand() - 0.5) * 20, 5, 90).round();

  final switchBase = 8 + totalStress * 27;
  final appSwitches = _clamp(switchBase + (rand() - 0.5) * 6, 5, 40).round();

  final sleepScore = _clamp((sleepHours - 4) / 5, 0, 1);
  final screenScore = _clamp(1 - (screenTimeHours - 1.5) / 7, 0, 1);
  final activityScore = _clamp((activeMinutes - 5) / 85, 0, 1);
  final focusScore = _clamp(1 - (appSwitches - 5) / 30, 0, 1);

  final rawWellness =
      (sleepScore * 0.30 + screenScore * 0.25 + activityScore * 0.20 + focusScore * 0.25) * 100;
  final wellnessScore = _clamp(rawWellness + (rand() - 0.5) * 6, 10, 98).round();

  return DailyData(
    date: dateStr,
    dayLabel: _dayNames[dow],
    dayOfWeek: dow,
    sleepHours: sleepHours,
    screenTimeHours: screenTimeHours,
    activeMinutes: activeMinutes,
    appSwitches: appSwitches,
    wellnessScore: wellnessScore,
  );
}

/// Public wrapper to generate synthetic data for an arbitrary date.
DailyData generateForDate(DateTime date) => _generateDay(date);

/// Compute focus score (0-100) from app switch count.
/// Higher switches = lower focus. Capped at 40 switches = 0%.
int computeFocusScore(int appSwitches) =>
    (100 - (appSwitches / 40 * 100)).round().clamp(0, 100);

List<DailyData> getWeeklyData({int days = 7}) {
  final now = DateTime.now();
  return List.generate(days, (i) {
    return _generateDay(now.subtract(Duration(days: days - 1 - i)));
  });
}

DailyData getTodayData() => _generateDay(DateTime.now());

List<BehavioralSignal> getTodaySignals({DailyData? todayOverride}) {
  final today = todayOverride ?? getTodayData();
  final yesterday = _generateDay(DateTime.now().subtract(const Duration(days: 1)));

  SignalTrend trend(double a, double b) {
    if (a > b) return SignalTrend.up;
    if (a < b) return SignalTrend.down;
    return SignalTrend.stable;
  }

  return [
    BehavioralSignal(
      id: 'sleep',
      label: 'Sleep',
      value: today.sleepHours.toStringAsFixed(1),
      unit: 'hours',
      trend: trend(today.sleepHours, yesterday.sleepHours),
      trendIsGood: today.sleepHours >= 7,
      icon: 'moon',
    ),
    BehavioralSignal(
      id: 'screen',
      label: 'Screen Time',
      value: today.screenTimeHours.toStringAsFixed(1),
      unit: 'hours',
      trend: trend(today.screenTimeHours, yesterday.screenTimeHours),
      trendIsGood: today.screenTimeHours <= 5,
      icon: 'phone',
    ),
    BehavioralSignal(
      id: 'movement',
      label: 'Active Minutes',
      value: today.activeMinutes.toString(),
      unit: 'min',
      trend: trend(today.activeMinutes.toDouble(), yesterday.activeMinutes.toDouble()),
      trendIsGood: today.activeMinutes >= 30,
      icon: 'walk',
    ),
    BehavioralSignal(
      id: 'focus',
      label: 'Focus Score',
      value: computeFocusScore(today.appSwitches).toString(),
      unit: '%',
      trend: trend(yesterday.appSwitches.toDouble(), today.appSwitches.toDouble()),
      trendIsGood: today.appSwitches <= 15,
      icon: 'target',
    ),
  ];
}


/// Enhanced today data that overlays real sensor data on the synthetic baseline.
///
/// Real values replace their synthetic counterparts, then the wellness score is
/// recalculated from the (possibly real) inputs.
DailyData getEnhancedTodayData({
  int? realSteps,
  int? moodRating,
  double? realSleepHours,
  double? realScreenTimeHours,
  int? realActiveMinutes,
  int? realAppCount,
}) {
  var base = getTodayData();

  // Overlay real sensor data onto synthetic baseline.
  base = base.copyWith(
    sleepHours: realSleepHours ?? base.sleepHours,
    screenTimeHours: realScreenTimeHours ?? base.screenTimeHours,
    activeMinutes: realActiveMinutes ?? base.activeMinutes,
    appSwitches: realAppCount ?? base.appSwitches,
    moodRating: moodRating,
    realSteps: realSteps,
  );

  final score = calculateWellnessScore(base, realSteps: realSteps);
  return base.copyWith(wellnessScore: score);
}

/// Calculate wellness score from the four behavioral signals.
///
/// Weights: sleep 0.30, screen time 0.25, activity 0.20, focus 0.25.
/// Optional [realSteps] and [DailyData.moodRating] add bonuses / penalties.
int calculateWellnessScore(DailyData data, {int? realSteps}) {
  final sleepScore = _clamp((data.sleepHours - 4) / 5, 0, 1);
  final screenScore = _clamp(1 - (data.screenTimeHours - 1.5) / 7, 0, 1);
  final activityScore = _clamp((data.activeMinutes - 5) / 85, 0, 1);
  final focusScore = _clamp(1 - (data.appSwitches - 5) / 30, 0, 1);

  double raw =
      (sleepScore * 0.30 + screenScore * 0.25 + activityScore * 0.20 + focusScore * 0.25) * 100;

  // Step bonus
  final steps = realSteps ?? data.realSteps;
  if (steps != null) {
    if (steps >= 8000) {
      raw += 5;
    } else if (steps >= 5000) {
      raw += 3;
    } else if (steps >= 2000) {
      raw += 1;
    } else {
      raw -= 3;
    }
  }

  // Mood adjustment
  if (data.moodRating != null) {
    raw += _moodAdjustment(data.moodRating!);
  }

  return raw.round().clamp(10, 98);
}

/// Returns a wellness score adjustment based on mood rating (1-5).
/// Mood 1 = -8, Mood 2 = -5, Mood 3 = 0, Mood 4 = +3, Mood 5 = +5
int _moodAdjustment(int moodRating) {
  switch (moodRating) {
    case 1:
      return -8;
    case 2:
      return -5;
    case 3:
      return 0;
    case 4:
      return 3;
    case 5:
      return 5;
    default:
      return 0;
  }
}

/// Detect anomalies by comparing today's data against the 7-day rolling average.
/// Returns a list of anomaly/warning objects for display in the UI.
///
/// If [history] is provided, it will be used for computing averages instead of
/// regenerating synthetic data. The last entry in [history] is treated as today.
/// If [history] is null, falls back to existing behavior using [getWeeklyData].
List<WellnessAnomaly> detectAnomalies({List<DailyData>? history}) {
  final week = history ?? getWeeklyData();
  if (week.isEmpty) return [];
  final today = week.last;
  final anomalies = <WellnessAnomaly>[];

  // Compute 7-day averages
  final avgSleep = week.map((d) => d.sleepHours).reduce((a, b) => a + b) / week.length;
  final avgScreen = week.map((d) => d.screenTimeHours).reduce((a, b) => a + b) / week.length;
  final avgWellness = week.map((d) => d.wellnessScore).reduce((a, b) => a + b) / week.length;
  final avgFocus =
      week.map((d) => computeFocusScore(d.appSwitches).toDouble()).reduce((a, b) => a + b) / week.length;

  // Sleep < 6h absolute warning
  if (today.sleepHours < 6) {
    anomalies.add(WellnessAnomaly(
      id: 'sleep-low',
      type: AnomalyType.warning,
      title: 'Low sleep detected',
      message:
          'You logged ${today.sleepHours}h of sleep. Aim for at least 7 hours to support recovery.',
      metric: 'sleep',
      severity: today.sleepHours < 5 ? AnomalySeverity.high : AnomalySeverity.medium,
    ));
  }

  // Sleep dropped > 1.5h from average
  if (avgSleep - today.sleepHours > 1.5) {
    anomalies.add(WellnessAnomaly(
      id: 'sleep-drop',
      type: AnomalyType.warning,
      title: 'Sleep dropped significantly',
      message:
          'Your sleep is ${_round1(avgSleep - today.sleepHours)}h below your weekly average of ${_round1(avgSleep)}h.',
      metric: 'sleep',
      severity: AnomalySeverity.medium,
    ));
  }

  // Screen time > 6h absolute warning
  if (today.screenTimeHours > 6) {
    anomalies.add(WellnessAnomaly(
      id: 'screen-high',
      type: AnomalyType.warning,
      title: 'High screen time',
      message:
          'Screen time is at ${today.screenTimeHours}h today. Consider taking a break from devices.',
      metric: 'screenTime',
      severity: today.screenTimeHours > 7.5 ? AnomalySeverity.high : AnomalySeverity.medium,
    ));
  }

  // Screen time jumped > 2h above average
  if (today.screenTimeHours - avgScreen > 2) {
    anomalies.add(WellnessAnomaly(
      id: 'screen-spike',
      type: AnomalyType.warning,
      title: 'Screen time spike',
      message:
          'Screen time is ${_round1(today.screenTimeHours - avgScreen)}h above your weekly average of ${_round1(avgScreen)}h.',
      metric: 'screenTime',
      severity: AnomalySeverity.medium,
    ));
  }

  // Wellness score < 50 (high warning)
  if (today.wellnessScore < 50) {
    anomalies.add(WellnessAnomaly(
      id: 'wellness-low',
      type: AnomalyType.warning,
      title: 'Wellness score is low',
      message:
          'Your score of ${today.wellnessScore} is below the healthy range. Focus on sleep and reducing screen time.',
      metric: 'wellness',
      severity: AnomalySeverity.high,
    ));
  }

  // Wellness improving trend (last 3 days rising)
  if (week.length >= 3) {
    final last3 = week.sublist(week.length - 3);
    final isImproving = last3[2].wellnessScore > last3[1].wellnessScore &&
        last3[1].wellnessScore > last3[0].wellnessScore;
    if (isImproving && today.wellnessScore > avgWellness) {
      anomalies.add(const WellnessAnomaly(
        id: 'wellness-improving',
        type: AnomalyType.positive,
        title: 'Wellness is trending up',
        message: 'Your score has been rising over the past 3 days. Keep up the good habits.',
        metric: 'wellness',
        severity: AnomalySeverity.low,
      ));
    }
  }

  // Low mood warning
  if (today.moodRating != null && today.moodRating! <= 2) {
    anomalies.add(WellnessAnomaly(
      id: 'mood-low',
      type: AnomalyType.warning,
      title: 'Low mood reported',
      message:
          'You rated your mood ${today.moodRating}/5. Consider a breathing exercise or reaching out to someone.',
      metric: 'mood',
      severity: today.moodRating == 1 ? AnomalySeverity.high : AnomalySeverity.medium,
    ));
  }

  // Focus score dropping (today vs average)
  final todayFocus = computeFocusScore(today.appSwitches).toDouble();
  if (avgFocus - todayFocus > 15) {
    anomalies.add(WellnessAnomaly(
      id: 'focus-drop',
      type: AnomalyType.info,
      title: 'Focus score is lower than usual',
      message:
          'Your focus is ${(avgFocus - todayFocus).round()} points below your weekly average. Try minimizing app switching.',
      metric: 'focus',
      severity: AnomalySeverity.low,
    ));
  }

  return anomalies;
}

/// Generate a smart insight that references step data, anomalies, and weekly trends.
/// Replaces [getWeeklyInsight] with richer, more contextual output.
///
/// If [history] is provided, it will be used instead of regenerating synthetic data.
/// The last entry in [history] is treated as today.
String getSmartInsight({int? realSteps, List<DailyData>? history}) {
  final week = history ?? getWeeklyData();
  if (week.isEmpty) return 'No data available yet.';
  final today = week.last;
  final avgSleep = week.map((d) => d.sleepHours).reduce((a, b) => a + b) / week.length;
  final avgScreen = week.map((d) => d.screenTimeHours).reduce((a, b) => a + b) / week.length;
  final worstDay = week.reduce((a, b) => a.wellnessScore < b.wellnessScore ? a : b);
  final bestDay = week.reduce((a, b) => a.wellnessScore > b.wellnessScore ? a : b);
  final scoreTrend = today.wellnessScore - week.first.wellnessScore;
  final anomalies = detectAnomalies(history: history);
  final warningCount = anomalies.where((a) => a.type == AnomalyType.warning).length;

  // Step-based context
  final steps = realSteps ?? today.realSteps;
  var stepContext = '';
  if (steps != null) {
    final formatted = _formatNumber(steps);
    if (steps >= 8000) {
      stepContext = " You've walked $formatted steps today — excellent activity level.";
    } else if (steps >= 5000) {
      stepContext = " With $formatted steps so far, you're moderately active. Try to reach 8,000.";
    } else if (steps >= 2000) {
      stepContext =
          " You're at $formatted steps — a short walk could boost your mood and focus.";
    } else {
      stepContext =
          ' Only $formatted steps recorded. Physical activity strongly correlates with better sleep and focus.';
    }
  }

  // Mood-based context
  var moodContext = '';
  if (today.moodRating != null) {
    if (today.moodRating! <= 2) {
      moodContext = ' Your self-reported mood is low — be gentle with yourself today.';
    } else if (today.moodRating! >= 4) {
      moodContext = ' Great to see your mood is positive today.';
    }
  }

  // Priority-ordered insight selection
  if (warningCount >= 2) {
    final metrics =
        anomalies.where((a) => a.type == AnomalyType.warning).map((a) => a.metric).join(', ');
    return 'Multiple areas need attention today: $metrics. Small improvements in any one area can lift your overall wellbeing.$moodContext$stepContext';
  }
  if (today.sleepHours < 6) {
    return 'You slept only ${today.sleepHours}h last night. Sleep deprivation compounds — even one recovery night helps. Consider winding down earlier tonight.$moodContext$stepContext';
  }
  if (today.screenTimeHours > 6) {
    return 'Screen time is elevated at ${today.screenTimeHours}h today. Try the 20-20-20 rule: every 20 minutes, look at something 20 feet away for 20 seconds.$moodContext$stepContext';
  }
  if (today.wellnessScore < 50) {
    return 'Your wellness score is ${today.wellnessScore} today. ${worstDay.dayLabel} was your toughest day this week at ${worstDay.wellnessScore}. Focus on one improvement — sleep or screen time — to recover.$moodContext$stepContext';
  }
  if (avgSleep < 6.5) {
    return 'Your average sleep this week is ${_round1(avgSleep)}h — below the recommended 7-9h. Sleep is the strongest predictor of next-day wellbeing in students.$moodContext$stepContext';
  }
  if (scoreTrend > 10) {
    return 'Your wellness trend is improving (+$scoreTrend points since ${week.first.dayLabel}). Keep maintaining your current routine — consistency is key.$moodContext$stepContext';
  }
  if (avgScreen > 5.5) {
    return 'Average screen time is ${_round1(avgScreen)}h this week. Consider setting app timers for your most-used apps to stay mindful of usage.$moodContext$stepContext';
  }
  if (anomalies.any((a) => a.type == AnomalyType.positive)) {
    return 'Great progress — your wellness has been improving steadily. ${bestDay.dayLabel} was your best day at ${bestDay.wellnessScore}. Keep up the momentum.$moodContext$stepContext';
  }
  return 'Your wellness has been ${today.wellnessScore >= 70 ? 'steady' : 'variable'} this week. ${bestDay.dayLabel} was your strongest day at ${bestDay.wellnessScore}. Keep prioritizing sleep and breaks.$moodContext$stepContext';
}

String _formatNumber(int n) {
  if (n < 1000) return n.toString();
  final thousands = n ~/ 1000;
  final remainder = n % 1000;
  if (remainder == 0) return '$thousands,000';
  return '$thousands,${remainder.toString().padLeft(3, '0')}';
}
